File: servidor/app/cantidades.py
"""Conversión entre texto y enteros escalados.

Las cantidades viajan y se guardan en milésimas, y los importes en
centavos. Sumar float acumula error, y un inventario suma miles de veces.
"""

import re
from decimal import ROUND_HALF_UP, Decimal, InvalidOperation

MILESIMAS = 1000
CENTAVOS = 100

SOLO_DIGITOS = re.compile(r"^-?\d+$")


def _agrupacion_valida(entero, separador_miles):
    """Verifica que los grupos de miles tengan tres dígitos.

    Sin esto «1,2,3» se leería como 123 en vez de rechazarse, y un número
    inventado es peor que un error: pasa por una cantidad real.
    """
    if separador_miles not in entero:
        return bool(SOLO_DIGITOS.match(entero))
    patron = r"^-?\d{1,3}(" + re.escape(separador_miles) + r"\d{3})+$"
    return bool(re.match(patron, entero))


def _partir(limpio, texto):
    """Separa parte entera y decimal resolviendo el formato del origen.

    Los ERP exportan indistintamente 1.234,56 y 1,234.56. Cuando conviven
    los dos caracteres, el último es el decimal y el otro agrupa miles.
    Cuando hay uno solo repetido (1.234.567) solo puede agrupar miles.

    Queda una ambigüedad que ningún criterio resuelve: «1.234» puede ser mil
    doscientos treinta y cuatro o uno coma doscientos treinta y cuatro. Se
    interpreta como decimal, que es lo habitual en cantidades. La vista
    previa de la importación existe para detectar el caso contrario.
    """
    coma = limpio.rfind(",")
    punto = limpio.rfind(".")

    if coma == -1 and punto == -1:
        if not SOLO_DIGITOS.match(limpio):
            raise ValueError(f"Cantidad inválida: {texto!r}")
        return limpio, ""

    if coma >= 0 and punto >= 0:
        separador_decimal = "," if coma > punto else "."
    else:
        unico = "," if coma >= 0 else "."
        if limpio.count(unico) > 1:
            if not _agrupacion_valida(limpio, unico):
                raise ValueError(f"Cantidad inválida: {texto!r}")
            return limpio.replace(unico, ""), ""
        separador_decimal = unico

    separador_miles = "." if separador_decimal == "," else ","
    entero, _, decimal = limpio.rpartition(separador_decimal)

    # Un separador suelto («.» o «,») no es cero: es una celda rota. Devolver
    # cero sería lo peor posible, porque un cero pasa por un conteo real.
    if not decimal.isdigit():
        raise ValueError(f"Cantidad inválida: {texto!r}")

    if entero in ("", "-"):
        entero += "0"
    if not _agrupacion_valida(entero, separador_miles):
        raise ValueError(f"Cantidad inválida: {texto!r}")

    return entero.replace(separador_miles, ""), decimal


def _a_escalado(texto, escala):
    if not isinstance(texto, str):
        raise ValueError(f"Se esperaba texto y llegó {type(texto).__name__}")

    limpio = texto.strip()
    if not limpio:
        raise ValueError("Cantidad vacía")

    entero, decimal = _partir(limpio, texto)

    try:
        valor = Decimal(f"{entero}.{decimal or 0}")
        # Redondeo y no truncamiento: el ERP puede exportar más decimales de
        # los que se guardan, y truncar sesgaría todas las cantidades a la baja.
        return int((valor * escala).quantize(Decimal(1), rounding=ROUND_HALF_UP))
    except (InvalidOperation, ArithmeticError) as error:
        raise ValueError(f"Cantidad inválida: {texto!r}") from error


def a_milesimas(texto: str) -> int:
    """Convierte texto a milésimas. '3,5' -> 3500."""
    return _a_escalado(texto, MILESIMAS)


def a_centavos(texto: str) -> int:
    """Convierte texto a centavos. '1.234,56' -> 123456."""
    return _a_escalado(texto, CENTAVOS)


def a_texto_importe(centavos: int | None) -> str:
    """Convierte centavos a texto con dos decimales. 12345 -> '123,45'.

    Vive acá, junto a `a_centavos`, para que la conversión de importes tenga
    una sola implementación: la exportación y el panel muestran plata los dos.
    A diferencia de `a_texto`, nunca recorta los decimales: un importe de
    '10' a secas se lee como diez pesos y no como diez con cero centavos.
    """
    if centavos is None:
        return ""
    signo = "-" if centavos < 0 else ""
    entero, resto = divmod(abs(centavos), CENTAVOS)
    return f"{signo}{entero},{resto:02d}"


def a_texto(milesimas: int) -> str:
    """Convierte milésimas a texto con coma decimal. 3500 -> '3,5'."""
    signo = "-" if milesimas < 0 else ""
    absoluto = abs(milesimas)
    entero, resto = divmod(absoluto, MILESIMAS)

    if resto == 0:
        return f"{signo}{entero}"

    decimales = f"{resto:03d}".rstrip("0")
    return f"{signo}{entero},{decimales}"

File: celular/nucleo/src/main/kotlin/com/controldestock/nucleo/Cantidades.kt
package com.controldestock.nucleo

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Conversión entre texto y enteros escalados.
 *
 * Las cantidades viajan y se guardan en milésimas. Sumar punto flotante
 * acumula error, y un inventario suma miles de veces.
 *
 * Es la contraparte exacta de `servidor/app/cantidades.py`: los dos lados
 * comparten la tabla de casos de prueba.
 */
object Cantidades {

    private const val MILESIMAS = 1000
    private val SOLO_DIGITOS = Regex("^-?\\d+$")

    fun aMilesimas(texto: String): Int {
        val limpio = texto.trim()
        require(limpio.isNotEmpty()) { "La cantidad está vacía" }

        val (entero, decimal) = partir(limpio)

        val valor = try {
            BigDecimal("$entero.${decimal.ifEmpty { "0" }}")
        } catch (error: NumberFormatException) {
            throw IllegalArgumentException("Cantidad inválida: «$texto»", error)
        }

        // Redondeo y no truncamiento: truncar sesgaría todas las cantidades a
        // la baja de forma sistemática.
        return valor.multiply(BigDecimal(MILESIMAS))
            .setScale(0, RoundingMode.HALF_UP)
            .toInt()
    }

    fun aTexto(milesimas: Int): String {
        val signo = if (milesimas < 0) "-" else ""
        val absoluto = Math.abs(milesimas)
        val entero = absoluto / MILESIMAS
        val resto = absoluto % MILESIMAS

        if (resto == 0) return "$signo$entero"

        val decimales = resto.toString().padStart(3, '0').trimEnd('0')
        return "$signo$entero,$decimales"
    }

    /**
     * Separa parte entera y decimal resolviendo el formato del origen.
     *
     * Los ERP exportan indistintamente 1.234,56 y 1,234.56. Cuando conviven
     * los dos caracteres, el último es el decimal y el otro agrupa miles.
     * Cuando hay uno solo repetido (1.234.567) solo puede agrupar miles.
     */
    private fun partir(limpio: String): Pair<String, String> {
        val coma = limpio.lastIndexOf(',')
        val punto = limpio.lastIndexOf('.')

        if (coma == -1 && punto == -1) {
            require(SOLO_DIGITOS.matches(limpio)) { "Cantidad inválida: «$limpio»" }
            return limpio to ""
        }

        val separadorDecimal: Char
        if (coma >= 0 && punto >= 0) {
            separadorDecimal = if (coma > punto) ',' else '.'
        } else {
            val unico = if (coma >= 0) ',' else '.'
            if (limpio.count { it == unico } > 1) {
                require(agrupacionValida(limpio, unico)) { "Cantidad inválida: «$limpio»" }
                return limpio.replace(unico.toString(), "") to ""
            }
            separadorDecimal = unico
        }

        val separadorMiles = if (separadorDecimal == ',') '.' else ','
        val corte = limpio.lastIndexOf(separadorDecimal)
        var entero = limpio.substring(0, corte)
        val decimal = limpio.substring(corte + 1)

        // Un separador suelto no es cero: es una celda rota. Devolver cero
        // sería lo peor posible, porque un cero pasa por un conteo real.
        require(decimal.isNotEmpty() && decimal.all { it.isDigit() }) {
            "Cantidad inválida: «$limpio»"
        }

        if (entero.isEmpty() || entero == "-") entero += "0"
        require(agrupacionValida(entero, separadorMiles)) { "Cantidad inválida: «$limpio»" }

        return entero.replace(separadorMiles.toString(), "") to decimal
    }

    /**
     * Verifica que los grupos de miles tengan tres dígitos.
     *
     * Sin esto «1,2,3» se leería como 123 en vez de rechazarse, y un número
     * inventado es peor que un error: pasa por una cantidad real.
     */
    private fun agrupacionValida(entero: String, separadorMiles: Char): Boolean {
        if (!entero.contains(separadorMiles)) return SOLO_DIGITOS.matches(entero)
        val escapado = Regex.escape(separadorMiles.toString())
        return Regex("^-?\\d{1,3}($escapado\\d{3})+$").matches(entero)
    }
}
